server/pulp/plugins/rsync/publish.py

```diff
@@ -0,0 +1,522 @@
+
+
+ASSOCIATED_UNIT_DATE_KEYWORD = "created"
+START_DATE_KEYWORD = 'start_date'
+END_DATE_KEYWORD = 'end_date'
+
+_logger = logging.getLogger(__name__)
+
+
+class RSyncPublishStep(PublishStep):
+    _CMD = "rsync"
+    # recursive, symlinks, timestamps, keep dir links, omit dir times, compress, itemize
+    _AZQ = "-rtKOzi"
+
+    def __init__(self, step_type, file_list, src_prefix, dest_prefix,
+                 working_dir=None, distributor_type=None, config=None,
+                 fast_forward=None, exclude=[], delete=False, links=False):
+        """
+        Copies files using rsync to a remote server and filesystem location defined in the
+        distributor config.
+
+        :param step_type: Name of step
+        :type step_type: str
+        :param file_list: list of files to rsync
+        :type file_list: list
+        :param src_prefix: path to directory in Pulp which contains all the files
+        :type src_prefix: str
+        :param dest_prefix: path to directory on remote server where files should be rsynced
+        :type dest_prefix: str
+        :param working_dir:
+        :param distributor_type: distributor type id
+        :param config: distributor configuration
+        :param fast_forward:
+        :param exclude: list of files/directories to skip
+        :type exclude: list
+        :param delete: determines if --delete is passed to rsync
+        :type delete: boolean
+        """
+        self.description = _('Rsync files to remote destination')
+        super(PublishStep, self).__init__(step_type, working_dir=working_dir, config=config)
+        self.file_list = file_list
+        self.fast_forward = fast_forward
+        self.exclude = exclude
+        self.delete = delete
+        self.src_prefix = src_prefix
+        self.dest_prefix = dest_prefix
+        self.links = links
+
+    def remote_mkdir(self, path):
+        """
+        Creates path on remote server. The path is rooted in distributor's remote_root directory.
+
+        :param path: path to create on remote server
+        :type path: str
+        :return: (return code from rsync, stdout + stderr)
+        :rtype: tuple
+        """
+        tmpdir = os.path.join(self.get_working_dir(), '.tmp')
+        os.makedirs(os.path.join(tmpdir, path.lstrip("/")))
+        args = ['rsync', '-avr', '-f+ */']
+        args.extend(self.make_authentication())
+        args.append("%s/" % tmpdir)
+        args.append(self.make_destination(path).replace(str(path), ""))
+        _logger.info("remote mkdir: %s" % args)
+        is_ok, output = self.call(args, include_args_in_output=True)
+        _logger.info("remote mkdir out: %s" % output)
+        shutil.rmtree(tmpdir)
+        return is_ok, output
+
+    def make_ssh_cmd(self, args=None):
+        """
+        Returns a list of arguments needed to form an ssh command for connecting to remote server.
+
+        :param args:
+        :type args: list of extra args to append to the standard ssh command
```

Comment:
looks like this text should be for the ``param`` line, and just ``list`` should be this line.